controllers/todo.js

```diff
@@ -1,40 +1,173 @@
+  const { title } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!title) {
+    res.status(400).send("Title field is empty!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const queryTodo = await ToDo.findByIdAndUpdate(todoId, { title: title });
+
+    if (queryTodo) {
+      res.status(200).json({ id: todoId, title: title });
+    } else {
+      res.status(404).send("Todo of given id not found!");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
 const editToDoPatch = async (req, res) => {
-  // Change the title of the Todo with given id, and get the new title as response
+  const { title } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!title) {
+    res.status(400).send("Title field is empty!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const queryTodo = await ToDo.findByIdAndUpdate(todoId, { title: title });
+
+    if (queryTodo) {
+      res.status(200).json({ id: todoId, title: title });
+    } else {
+      res.status(404).send("Todo of given id not found!");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
 const deleteToDo = async (req, res) => {
-  //  Delete the todo with given id
+  const todoId = req.params.id;
+  const userId = req.user.id;
+  const todo = await ToDo.findById(todoId);
+
+  if (!todo) {
+    res.status(404).send("Todo of given id not found!");
+  } else if (creator(userId, todo) || collaborator(userId, todo)) {
+    const deleteTodo = await ToDo.findByIdAndDelete(todoId);
+
+    if (deleteTodo) {
+      res.status(204).send("Todo of given id deleted");
+    } else {
+      res.status(501).send("Couldn't delete task.");
+    }
+  } else {
+    res.status(403).send("Logged user is not authorized to edit todo of given id!");
+  }
 };
 
+const addCollaborator = async (req, res) => {
+  const { collaboratorId } = req.body;
+  const todoId = req.params.id;
+  const userId = req.user.id;
+
+  const collaborator = await User.findById(collaboratorId);
+  const todo = await ToDo.findById(todoId);
+
+  const flag = collaboratorId != todo.createdBy;
+
+  if (creator(userId, todo) && flag) {
+    collaborator.collaboratingTodos.push(todoId);
+    todo.collaborators.push(collaboratorId);
+    collaborator.save();
+    todo.save();
+    
+    res.status(200).json(todo);
+  } else if (!flag) {
+    res.status(400).send("Creater cannot collaborate!");
+  } else if (!creator(userId, todo)) {
+    res.status(403).send("Only creator of a todo can add collaborators");
+  } else {
+    res.status(404).send("User or Todo of given id not found!");
+  }
+}
+
+const removeCollaborator = async (req, res) => {
```

Comment:
Try adding a case where if collaborator is not present then I get the proper message.

Reply:
okay. nice point, i missed it.